### DISCUSSION TITLE
Inclusion of <nlohmann/json.hpp> imprints header path to binary, destroying path-independent reproductive builds

### DISCUSSION THREAD

**Discussion by usvi** [nlohmann/json#4494] [Q&A] [answered]
Hello,

Whenever I do #include <nlohmann/json.hpp> the library imprints the full path of the header file to the binary. This essentially destroys path-independent builds, for example in Yocto. How do I configure or modify the library NOT to print anything as this path?

**EDIT1**:
Version is 2.1.1

**EDIT2**:

Minimal test case:

main.cpp:

```
#include <stdio.h>
#include <nlohmann/json.hpp>

int main(int i_argc, char** as_argv)
{
  nlohmann::json x_json;

  printf("Hello Yocto World!\n");

  return 0;
}
```
```
user@MACHINE:~/repos/helloyoctoworld$ g++ -o test main.cpp
user@MACHINE:~/repos/helloyoctoworld$ strings test | grep hpp
/usr/include/nlohmann/json.hpp
user@MACHINE:~/repos/helloyoctoworld$
```

**Comment by nlohmann** [comment]
Can this be an issue with your compiler/linker setup?

**Reply by usvi** [reply]
I really don't know for sure but I doubt it. I'm also doing #include <mosquitto.h> but it does not imprint anything. My initial hunch is that json.hpp caches the path of the header file to use in some kind of debug prints or something. Could it be? Is there debug printing properties using the header name? I tested adding the header to multiple cpp files. Every inclusion = new header path in the final binary.

Can I do something to investigate?

**Reply by nlohmann** [reply]
I tried `g++ -E single_include/nlohmann/json.hpp` but did not find an absolute path to the `json.hpp` file in the output. I also called `strings` on some test binaries and also did not see a path there.

**Reply by usvi** [reply]
I made a simple test case in the beginning. Am I compiling it wrong? I'm using 2.1.1 which ships with Ubuntu 18.04.6 LTS. 

```
user@MACHINE:~/repos/helloyoctoworld$ g++ --version
g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by usvi** [comment]
I made a self-contained test case:

https://github.com/usvi/nlohmannjsontest

See the output:

janne@shell:/tmp$  git clone git@github.com:usvi/nlohmannjsontest.git
Cloning into 'nlohmannjsontest'...
remote: Enumerating objects: 18, done.
remote: Counting objects: 100% (18/18), done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 18 (delta 2), reused 18 (delta 2), pack-reused 0 (from 0)
Receiving objects: 100% (18/18), 204.74 KiB | 605.00 KiB/s, done.
Resolving deltas: 100% (2/2), done.
janne@shell:/tmp$ cd nlohmannjsontest/
janne@shell:/tmp/nlohmannjsontest$ make
g++  -I inc -Wno-deprecated  -c -o main_3.11.3.o src/main_3.11.3.cpp
g++   main_3.11.3.o   -o main_3.11.3
g++  -I inc -Wno-deprecated  -c -o main_2.1.1.o src/main_2.1.1.cpp
g++   main_2.1.1.o   -o main_2.1.1
strip main_3.11.3
strip main_2.1.1
strings main_3.11.3 | grep "json.hpp"
inc/nlohmann3.11.3/json.hpp
strings main_2.1.1 | grep "json.hpp"
inc/nlohmann2.1.1/json.hpp

So, the strings are inside the binaries. What can I do?

**Reply by gregmarr** [reply, marked as answer]
It's probably from `assert`s.  Try including `-DNDEBUG` and `-O2`.

**Reply by usvi** [reply]
Thank you! Making #define NDEBUG before header inclusion and undeffing it afterwards has taken the assert path away. I am quite confident that the builds will now be reproductible with different paths.

**Reply by gregmarr** [reply]
If you just want to get rid of asserts, you can add this before the include.
```
#define JSON_ASSERT(x)
```

**Comment by nlohmann** [comment]
I extended the README, see #4496.

**Reply by usvi** [reply]
Thank you!
